src/database-controller/src/watcher/cluster-event/index.js

```diff
@@ -0,0 +1,139 @@
+const logger = require('@dbc/common/logger');
+const { getEventInformer } = require('@dbc/common/k8s');
+const { alwaysRetryDecorator } = require('@dbc/common/util');
+const disk = require('diskusage');
+const config = require('@dbc/watcher/cluster-event/config');
+
+// Here, we use AsyncLock to control the concurrency of events with the same uid;
+// e.g. If one event has ADDED, MODIFED, and MODIFED incidents, we use AsyncLock
+// to ensure they will be delivered to write-merger in order.
+// In the same time, we use PQueue to control the concurrency of events with different uid;
+// e.g. If there are event 1 ~ event 30000, only some of them can be processed concurrently.
+const lock = new AsyncLock({ maxPending: Number.MAX_SAFE_INTEGER });
+const queue = new PQueue({ concurrency: config.maxRpcConcurrency });
+const databaseModel = new DatabaseModel(
+  config.dbConnectionStr,
+  config.maxDatabaseConnection,
+);
+
+async function synchronizeEvent(eventType, apiObject) {
+  // query db instead
+  const uid = apiObject.metadata.uid;
+  const names = apiObject.involvedObject.name.split('-');
+
+  const obj = {
+    uid: uid,
+    frameworkName: names[0],
+    podUid: apiObject.involvedObject.uid,
+    taskroleName: names[1],
+    taskName: apiObject.involvedObject.name,
+    taskIndex: parseInt(names[2]),
+    type: apiObject.type,
+    reason: apiObject.reason,
+    message: apiObject.message,
+    firstTimestamp: apiObject.firstTimestamp,
+    lastTimestamp: apiObject.lastTimestamp,
+    count: apiObject.count,
+    sourceComponent: _.get(apiObject, 'source.component', null),
+    sourceHost: _.get(apiObject, 'source.host', null),
+    event: JSON.stringify(apiObject),
+  };
+
+  databaseModel.FrameworkEvent.upsert(obj, { where: { uid: uid } });
+}
+
+const eventHandler = (eventType, apiObject) => {
+  /*
+    framework name-based lock + always retry
+  */
+  const receivedTs = new Date().getTime();
+  const involvedObjKind = apiObject.involvedObject.kind;
+  const involvedObjName = apiObject.involvedObject.name;
+  const uid = apiObject.metadata.uid;
+  if (
+    involvedObjKind === 'Pod' &&
+    /^[a-z0-9]{32}-[A-Za-z0-9._~]+-[0-9]+$/.test(involvedObjName)
+  ) {
+    logger.info(
+      `Cluster event type=${eventType} receivedTs=${receivedTs} uid=${uid} involvedObjKind=${involvedObjKind} involvedObjName=${involvedObjName} received.`,
+    );
+    lock.acquire(uid, () => {
+      return queue.add(
+        alwaysRetryDecorator(
+          () => synchronizeEvent(eventType, apiObject),
+          `Sync to database type=${eventType} receivedTs=${receivedTs} uid=${uid} involvedObjKind=${involvedObjKind} involvedObjName=${involvedObjName}`,
+        ),
+      );
+    });
+  } else {
+    logger.info(
+      `Cluster Event type=${eventType} receivedTs=${receivedTs} uid=${uid} involvedObjKind=${involvedObjKind} involvedObjName=${involvedObjName} received but ignored.`,
+    );
+  }
+};
+
+async function assertDiskUsageHealthy() {
```

Comment:
Better to limit the quota per job and global, and does not impact the critical path.

Reply:
Recorded in #4953 . We can solve this problem in the future.